api/src/services/export/dhis.js

```diff
@@ -0,0 +1,215 @@
+  const dhisTargets = settingsDoc.tasks &&
+    settingsDoc.tasks.targets &&
+    settingsDoc.tasks.targets.items &&
+    settingsDoc.tasks.targets.items.filter(target =>
+      target.dhis &&
+      target.dhis.dataElement &&
+      (!target.dhis.dataSet || target.dhis.dataSet === dataSet)
+    ) || [];
+
+  return dhisTargets;
+};
+
+/**
+ * @param {string} dataSet The dataset being exported. It acts as a filter for relevant orgUnits
+ * @param {Object[]} contacts The set of contact documents relevant to the calculation
+ * @returns {Object} The @param contacts _id values mapped to an array of orgUnits above them in the hierarchy
+ */
+const mapContactIdToOrgUnits = (dataSet, contacts, contactsWithOrgUnits) => {
+  const result = {};
+  for (const contact of contactsWithOrgUnits) {
+    const dhisConfigs = Array.isArray(contact.dhis) ? contact.dhis : [contact.dhis];
+    for (const dhisConfig of dhisConfigs) {
+      const dataSetMatch = !dhisConfig.dataSet || dhisConfig.dataSet === dataSet;
+      if (dhisConfig.orgUnit && dataSetMatch) {
+        if (!result[contact._id]) {
+          result[contact._id] = [];
+        }
+
+        result[contact._id].push(dhisConfig.orgUnit);
+      }
+    }
+  }
+  
+  for (const contact of contacts) {
+    let traverse = contact;
+    while (traverse) {
+      if (result[traverse._id]) {
+        const existing = result[contact._id] || [];
+        result[contact._id] = _.uniq([...existing, ...result[traverse._id]]);
+      }
+
+      traverse = traverse.parent;
+    }
+  }
+
+  return result;
+};
+
+const buildDataValues = (targetDefinitions, targetDocs, orgUnits) => {
+  const mapTargetIdToDhis = ObjectFromEntries(targetDefinitions.map(target => ([target.id, target.dhis])));
+  const createEmptyValueSetFor = orgUnit => {
+    const result = {};
+    for (const target of targetDefinitions) {
+      const { dataElement } = target.dhis;
+      result[dataElement] = Object.assign(
+        {},
+
+        /*
+        Copies any attribute defined in dhis config onto the dataValue
+        Primary usecase is `categoryOptionCombo` and `attributeOptionCombo` but there are many others
+        */
+        target.dhis,
+        { orgUnit, value: 0 },
+      );
+      delete result[dataElement].dataSet;
+    }
+    return result;
+  };
+
+  // all results start with 0s
+  const dataValueSet = {};
+  for (const contactOrgUnits of Object.values(orgUnits)) {
+    for (const orgUnit of contactOrgUnits) {
+      if (!dataValueSet[orgUnit]) {
+        dataValueSet[orgUnit] = createEmptyValueSetFor(orgUnit);
+      }
+    }
+  }
+
+  // add relevant values onto the result
+  for (const targetDoc of targetDocs) {
+    const unitsOfOwner = orgUnits[targetDoc.owner];
+    if (!unitsOfOwner) {
+      continue;
+    }
+
+    for (const orgUnit of unitsOfOwner) {
+      for (const target of targetDoc.targets) {
+        const dataElement = mapTargetIdToDhis[target.id] && mapTargetIdToDhis[target.id].dataElement;
```

Comment:
I think these loops should be nested the other way around: this line isn't affected by `orgUnit`, but if it's falsy it blocks any need to loop the `orgUnit`.

So it would make more sense to loop through targets, work out the dataElement for that target and if it's truthy loop through the orgUnits for the owner.